### PR TITLE
Add S3 TransferException to track failed file source

### PR DESCRIPTION
**Pull request by cjyclaire**
Feature request from #1015.
By providing file information, customer would have better troubleshooting experience.

Besides having `getFileSouce()` to get the source file path that cause error, also add `getUncompletedFiles()` to retrieve remaining files that haven't been transferred.

cc:\ @mtdowling @xibz

### CODE REVIEW

File: src/S3/Transfer.php [hunk 1]
```diff
@@ -233,18 +239,30 @@ private function createDownloadPromise()
...
         }
 
         // Create a GetObject command pool and return the promise.
         return (new Aws\CommandPool($this->client, $commands, [
             'concurrency' => $this->concurrency,
             'before'      => $this->before,
+            'fulfilled'   => function ($value, $idx, Promise\PromiseInterface $p) {
+                if ($value !== null && $value instanceof Aws\ResultInterface) {
+                    $uri = $value->get('@metadata')['effectiveUri'];
+                    $this->clearTaskDone((new Uri($uri))->getPath());
```
Comment (mtdowling) [inline src/S3/Transfer.php]: This would then be updated to just remove the object from the SplObjectStorage

File: src/S3/Transfer.php [hunk 2]
```diff
@@ -233,18 +239,30 @@ private function createDownloadPromise()
...
+                }
+            },
             'rejected'    => function ($reason, $idx, Promise\PromiseInterface $p) {
-                $p->reject($reason);
+                $cmd = $reason->getCommand();
+                $this->handleTransferException(
```
Comment (mtdowling) [inline src/S3/Transfer.php]: I don't think we should be changing the exception here. All we need to do here is to remove the command from the inFlightTransfers object.

File: src/S3/Transfer.php [hunk 3]
```diff
@@ -233,18 +239,30 @@ private function createDownloadPromise()
...
+                    $reason,
+                    "/" . $cmd['Bucket'] . "/" . $cmd['Key']
+                );
             }
         ]))->promise();
```
Comment (mtdowling) [inline src/S3/Transfer.php]: This is where you'd wrap the exception by adding a `then` method on the `promise()` call that decorates the given exception, adds the information about the failed command, and the list of inFlightTransfers / remainingTransfers.

**Comment by cjyclaire** [comment]
We have talked about the PR offline, closing, tracking with #1140 instead

**Pull request closed without merging**